crates/ruff_db/src/diagnostic.rs

```diff
@@ -0,0 +1,175 @@
+use ruff_text_size::TextRange;
+use salsa::Accumulator as _;
+
+use crate::{
+    files::File,
+    source::{line_index, source_text},
+    Db,
+};
+
+pub trait Diagnostic: Send + Sync + std::fmt::Debug {
+    fn rule(&self) -> &str;
+
+    fn message(&self) -> std::borrow::Cow<str>;
+
+    fn file(&self) -> File;
+
+    fn range(&self) -> Option<TextRange>;
+
+    fn severity(&self) -> Severity;
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Severity {
+    Info,
+    Error,
+}
+
+#[salsa::accumulator]
+pub struct CompileDiagnostic(std::sync::Arc<dyn Diagnostic>);
+
+impl CompileDiagnostic {
+    pub fn report<T>(db: &dyn Db, diagnostic: T)
+    where
+        T: Diagnostic + 'static,
+    {
+        Self(std::sync::Arc::new(diagnostic)).accumulate(db);
+    }
+
+    pub fn rule(&self) -> &str {
+        self.0.rule()
+    }
+
+    pub fn message(&self) -> std::borrow::Cow<str> {
+        self.0.message()
+    }
+
+    pub fn file(&self) -> File {
+        self.0.file()
+    }
+
+    pub fn range(&self) -> Option<TextRange> {
+        self.0.range()
+    }
+
+    pub fn severity(&self) -> Severity {
+        self.0.severity()
+    }
```

Comment:
What's the benefit of implementing these methods in the `impl` block here, and then again in the trait `impl` directly below? Couldn't we just inline these definitions in the trait `impl`? I remember you telling me off for doing something like this back in https://github.com/astral-sh/ruff/pull/11564#discussion_r1615996481 ;)

Reply:
I've to be careful with you. You never forget lol.... 

I don't have a good answer. I was mainly annoyed that I had to bring the `Diagnostic` trait into scope everywhere were I sued the `CompileDiagnostic`. Looking at it now, it's exactly two places where this avoids importing the trait. I think it's fine to remove this for now